pkg/helm/release/manager_test.go

```diff
@@ -15,126 +15,80 @@
 package release
 
 import (
-	"encoding/json"
 	"testing"
 
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
 	"github.com/stretchr/testify/assert"
-	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
 )
 
-func newTestDeployment(containers []interface{}) *unstructured.Unstructured {
-	return &unstructured.Unstructured{
-		Object: map[string]interface{}{
-			"kind":       "Deployment",
-			"apiVersion": "apps/v1",
-			"metadata": map[string]interface{}{
-				"name":      "test",
-				"namespace": "ns",
-			},
-			"spec": map[string]interface{}{
-				"template": map[string]interface{}{
-					"spec": map[string]interface{}{
-						"containers": containers,
-					},
+func newTestDeployment(containers []v1.Container) *appsv1.Deployment {
+	return &appsv1.Deployment{
+		TypeMeta:   metav1.TypeMeta{Kind: "Deployment", APIVersion: "apps/v1"},
+		ObjectMeta: metav1.ObjectMeta{Name: "test", Namespace: "ns"},
+		Spec: appsv1.DeploymentSpec{
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: containers,
 				},
 			},
 		},
 	}
 }
 
-func TestManagerGeneratePatch(t *testing.T) {
+func TestManagerGenerateStrategicMergePatch(t *testing.T) {
 
 	tests := []struct {
-		o1    *unstructured.Unstructured
-		o2    *unstructured.Unstructured
-		patch []map[string]interface{}
+		o1    runtime.Object
+		o2    runtime.Object
+		patch string
 	}{
 		{
-			o1: newTestDeployment([]interface{}{
-				map[string]interface{}{
-					"name": "test1",
-				},
-				map[string]interface{}{
-					"name": "test2",
-				},
+			o1: newTestDeployment([]v1.Container{
+				{Name: "test1"},
+				{Name: "test2"},
 			}),
-			o2: newTestDeployment([]interface{}{
-				map[string]interface{}{
-					"name": "test1",
-				},
+			o2: newTestDeployment([]v1.Container{
+				{Name: "test1"},
 			}),
-			patch: []map[string]interface{}{},
+			patch: `{"spec":{"template":{"spec":{"$setElementOrder/containers":[{"name":"test1"}]}}}}`, //nolint:lll
```

Comment:
Is this the expected behavior? We are now removing a manual addition from the array. This would actually break my use of operator: A second operator(jaeger) is adding an additional sidecar container to all my deployments which are also managed the helm-operator.

I added a new testcase to test the behavior with new object attributes, this is handled correctly:
```
		{
			o1: &appsv1.Deployment{
				TypeMeta:   metav1.TypeMeta{Kind: "Deployment", APIVersion: "apps/v1"},
				ObjectMeta: metav1.ObjectMeta{
					Name: "test",
					Namespace: "ns",
					Annotations: map[string]string{
						"testannotation": "testvalue",
					},
				},
				Spec: appsv1.DeploymentSpec{},
			},
			o2: &appsv1.Deployment{
				TypeMeta:   metav1.TypeMeta{Kind: "Deployment", APIVersion: "apps/v1"},
				ObjectMeta: metav1.ObjectMeta{
					Name: "test",
					Namespace: "ns",
				},
				Spec: appsv1.DeploymentSpec{},
			},
			patch: `{}`, //nolint:lll
		},
```

Reply:
Hmm what exactly does this patch do? Does it even remove the second container? Or does it only enforce that the `test1` container is the first item in the array?

I haven't actually tested this in a cluster.

Reply:
In my actual testing, it does nothing.

I think since the patch strategy of `containers` is `merge`, there is no case where an item in the list will be outright removed.

I found some docs [here](https://stupefied-goodall-e282f7.netlify.com/contributors/design-proposals/cli/preserve-order-in-strategic-merge-patch/#when-the-list-to-be-merged-contains-elements-not-found-in-setelementorder) that describe that case, but they didn't line up with what I was actually seeing.

So I'm thinking that doc (which was a proposal) may not be what the final implementation was.